runtime/vm/ContinuationHelpers.hpp

```diff
@@ -67,14 +67,12 @@ class VM_ContinuationHelpers {
 		SWAP_MEMBER(stackOverflowMark, UDATA*, vmThread, continuation);
 		SWAP_MEMBER(stackOverflowMark2, UDATA*, vmThread, continuation);
 		SWAP_MEMBER(stackObject, J9JavaStack*, vmThread, continuation);
-	}
 
-	static VMINLINE void
-	popAndStoreELS(J9VMThread *vmThread, J9VMContinuation *continuation)
-	{
-		continuation->entryLocalStorage = *vmThread->entryLocalStorage;
-		continuation->entryLocalStorage.oldEntryLocalStorage = NULL;
-		vmThread->entryLocalStorage = vmThread->entryLocalStorage->oldEntryLocalStorage;
+		/* Swap the JIT GPR registers data referenced by ELS */
+		J9JITGPRSpillArea tempGPRs = continuation->jitGPRs;
+		continuation->jitGPRs = *(J9JITGPRSpillArea*)vmThread->entryLocalStorage->jitGlobalStorageBase;
+		*(J9JITGPRSpillArea*)vmThread->entryLocalStorage->jitGlobalStorageBase = tempGPRs;
+		SWAP_MEMBER(oldEntryLocalStorage, J9VMEntryLocalStorage*, vmThread->entryLocalStorage, continuation);
```

Comment:
The ELS pointer in the J9VMThread should never be modified - only the GPR spill area should be modified.

Reply:
I thought we agreed that the oldELS pointer should be saved and set to null when entering a continuation, and restored when yielding back to carrier

Reply:
So I should leave the oldELS on the vmThread regardless of continuation/carrier? and only update the J9VMContinuation data

Reply:
Correct.

Reply:
I suppose we could swap the oldELS pointer to be completely correct (so when the continuation is mounted, there is no previous).

Reply:
So keep the code you have, but swap the previous pointer, not the pointer in the J9VMThread.

Reply:
this is swapping the old ELS

Reply:
You may be able to use the swap macro on the spill area as well.

Reply:
> this is swapping the old ELS

No, it's swapping vmThread->entryLocalStorage which is the current ELS.

Reply:
`SWAP_MEMBER(oldEntryLocalStorage, J9VMEntryLocalStorage*, vmThread->entryLocalStorage, continuation);`

That's writing to the current ELS.

Reply:
Can you elaborate on this, I am a bit confused by

> `SWAP_MEMBER(oldEntryLocalStorage, J9VMEntryLocalStorage*, vmThread->entryLocalStorage, continuation);`
> 
> That's writing to the current ELS.

since the oldELS is stored in vmThread->entryLocalStorage->oldEntryLocalStorage, I don't see any other way to update that.

Reply:
Sorry, confused by the macro implementation - this is correct.